### PR TITLE
Propagating the UserErrors, UserWarnings, and PackageError variants

### PR DESCRIPTION
**Pull request by Alfred-Mountfield**
## 🌟 What is the purpose of this PR?

We distinguish between errors/warnings that come from the Runner and those that come from user-code. There was a TODO about implementing these variants and properly propagating them. This PR does that.

It also fixes a mistake where certain classifications of errors were marked as fatal for the runners (thus killing an experiment rather than just a simulation), when they shouldn't have been. (User errors and Package errors are non-fatal).

## 🔍 What does this change?

- Makes the msg strings on `PackageError` and `UserError` required in the FlatBuffers schemas. It doesn't make sense to propagate these without any associated error information
- Creates the `UserError`, `UserWarning`, and `PackageError` variants on all the relevant enums within the Rust code, propagating it up to the orchestrator for display.
- `UserError`s and `PackageError`s are matched against and no longer cause fatal exits of the Runners

### 📜 Does this require a change to the docs?

- No

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201707629991362/1201770186881429/f) (_internal_)

## 🛡 What tests cover this?

- None at the moment, we don't really test our reporting setup and/or failure conditions

### CODE REVIEW

File: packages/engine/src/worker/runner/comms/outbound.rs [hunk 1]
```diff
@@ -57,6 +58,60 @@ impl From<flatbuffers_gen::runner_warning_generated::RunnerWarning<'_>> for Runn
...
+
+impl From<flatbuffers_gen::user_error_generated::UserError<'_>> for UserError {
+    fn from(user_error: flatbuffers_gen::user_error_generated::UserError<'_>) -> Self {
+        Self(user_error.msg().to_string())
+    }
+}
+
+impl std::fmt::Display for UserError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
```
Comment (TimDiekmann) [inline packages/engine/src/worker/runner/comms/outbound.rs:72]: Let's forward to the display implementation.
```suggestion
        std::fmt::Display::fmt(&self.0, f)
```

File: packages/engine/src/worker/runner/comms/outbound.rs [hunk 2]
```diff
@@ -57,6 +58,60 @@ impl From<flatbuffers_gen::runner_warning_generated::RunnerWarning<'_>> for Runn
...
+
+impl From<flatbuffers_gen::package_error_generated::PackageError<'_>> for PackageError {
+    fn from(package_error: flatbuffers_gen::package_error_generated::PackageError<'_>) -> Self {
+        Self(package_error.msg().to_string())
+    }
+}
+
+impl std::fmt::Display for PackageError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.0)
```
Comment (TimDiekmann) [inline packages/engine/src/worker/runner/comms/outbound.rs:112]:
```suggestion
        std::fmt::Display::fmt(&self.0, f)
```

File: packages/engine/src/worker/runner/javascript/mod.rs [hunk 3]
```diff
@@ -378,38 +382,53 @@ fn schema_to_stream_bytes(schema: &Schema) -> Vec<u8> {
     stream_bytes
 }
 
-fn array_to_errors(array: mv8::Value<'_>) -> Vec<RunnerError> {
-    // TODO: Extract optional line numbers
+fn array_to_user_errors(array: mv8::Value<'_>) -> Vec<UserError> {
```
Comment (TimDiekmann) [inline packages/engine/src/worker/runner/javascript/mod.rs:385]: Don't we want to extract the line numbers here as well?

File: packages/engine/src/worker/runner/javascript/mod.rs [hunk 4]
```diff
@@ -909,7 +928,7 @@ impl<'m> RunnerImpl<'m> {
             payload_str.clone(),
         ]);
 
-        let (next_task_msg, warnings, logs) = self.run_task(
+        match self.run_task(
```
Comment (TimDiekmann) [inline packages/engine/src/worker/runner/javascript/mod.rs:931]: Can we write to a temporary variable (e.g. `run_result`) and `match run_result`? IMO this is more readable.

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

File: packages/engine/src/worker/runner/comms/outbound.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/engine/src/worker/runner/comms/outbound.rs:72]: I don't really feel strongly about this, so I'm happy to change, but the `write!` method is even the [one used in the rust-by-example docs](https://doc.rust-lang.org/rust-by-example/hello/print/print_display.html)

File: packages/engine/src/worker/runner/javascript/mod.rs (hunk 3 above)
Comment (Alfred-Mountfield) [reply packages/engine/src/worker/runner/javascript/mod.rs:385]: There isn't a line number field on the UserError message.
```
// fields:
//    `msg` : the error message
table UserError {
  msg:string (required);
}
```

File: packages/engine/src/worker/runner/comms/outbound.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/engine/src/worker/runner/comms/outbound.rs:72]: Also having a look through our code we much more frequently use `write!` in `Display` implementations so I'm actually going to leave it for now. If you feel strongly, or find an argument to support one over the other, lets open a ticket to change the code-base to be consistent, and then somehow capture it in some styling guidance

File: packages/engine/src/worker/runner/javascript/mod.rs (hunk 4 above)
Comment (Alfred-Mountfield) [reply packages/engine/src/worker/runner/javascript/mod.rs:931]: Done

**Review by TimDiekmann — APPROVED** [review]

**Pull request merged**